File: 0859-buddy-strings/0859-buddy-strings.java
class Solution {
    public boolean buddyStrings(String s, String goal) {
        if(s.length() != goal.length()) return false;
        if(s.equals(goal)) {
            int []freq = new int[26];
            for(int i=0; i<s.length(); i++) {
                freq[s.charAt(i) - 'a']++;
            }
            for(int i=0; i<26; i++) {
                if(freq[i] >= 2) return true;
            }
        }
        else {
            int x = -1;
            for(int i=0; i<s.length(); i++) {
                if(s.charAt(i) != goal.charAt(i)) {
                    if(x == -1) x = i;
                    else {
                        StringBuilder sb = new StringBuilder(s);
                        sb.setCharAt(x, s.charAt(i));
                        sb.setCharAt(i, s.charAt(x));
                        return goal.equals(sb.toString());
                    }
                }
            }
        }
        return false;
    }
}

File: 0859-buddy-strings/0859-buddy-strings.cpp
class Solution {
public:
    bool buddyStrings(string s, string goal) {
        if(s.size() != goal.size()) return false;
        if(s == goal) {
            vector<int>freq(26, 0);
            for(char c:s) freq[c - 'a']++;
            for(int i=0; i<26; i++) {
                if(freq[i] >= 2) return true;
            }
        }
        else {
            int x = -1;
            for(int i=0; i<s.size(); i++) {
                if(s[i] != goal[i]) {
                    if(x == -1) x = i;
                    else {
                        swap(s[i], s[x]);
                        return s == goal;
                    }
                }
            }
        }
        return false;
    }
};